client/src/includes/panels.ts

```diff
@@ -0,0 +1,82 @@
+/**
+ * Make panels collapsible, and collapse panels already marked as `collapsed`.
+ */
+export function initCollapsiblePanels() {
+  const toggles = document.querySelectorAll<HTMLButtonElement>(
+    '[data-panel-toggle]',
+  );
+
+  toggles.forEach((toggle) => {
+    const panel = toggle.closest<HTMLElement>('[data-panel]');
+    const content = document.querySelector<HTMLDivElement>(
+      `#${toggle.getAttribute('aria-controls')}`,
+    );
+
+    if (!content || !panel) {
+      return;
+    }
+
+    const onAnimationComplete = () => {
+      content.dispatchEvent(
+        new CustomEvent('commentAnchorVisibilityChange', { bubbles: true }),
+      );
+    };
+
+    const hasCollapsed = panel.classList.contains('collapsed');
+    const hasError = content.querySelector('[aria-invalid="true"]');
+
+    // Collapse panels marked as `collapsed`, unless they contain invalid fields.
+    if (hasCollapsed && !hasError) {
+      // Use experimental `until-found` value.
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      content.hidden = 'until-found';
+      toggle.setAttribute('aria-expanded', 'false');
+      onAnimationComplete();
+    }
+
+    toggle.addEventListener('click', () => {
+      const wasExpanded = toggle.getAttribute('aria-expanded') === 'true';
+      const isExpanded = !wasExpanded;
+      // Use experimental `until-found` value, so users can search inside the panels.
+      // Browsers without support for `until-found` will ignore the value.
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
+      content.hidden = !isExpanded ? 'until-found' : '';
+      onAnimationComplete();
+
+      toggle.setAttribute('aria-expanded', `${isExpanded}`);
+    });
+
+    const heading = panel.querySelector<HTMLHeadingElement>(
+      '[data-panel-heading]',
+    );
+
+    if (heading) {
+      heading.addEventListener('click', () => {
+        // When the section is closed, clicking the heading reveals it.
+        toggle.setAttribute('aria-expanded', 'true');
+        content.hidden = false;
+        onAnimationComplete();
+      });
+    }
+
+    // Set the toggle back to expanded upon reveal.
+    content.addEventListener('beforematch', () => {
+      toggle.setAttribute('aria-expanded', 'true');
+      onAnimationComplete();
+    });
```

Comment:
Can we also add an event listener for custom triggering

```
/* allow custom triggering */
panel.addEventListener('wagtail:panel-toggle', /*...*/);
```

AND for dispatching when triggered / open & closed.

```
panel.dispatch(... `wagtail:panel-open` etc
```

Reply:
Spoke with Thibaud

* Dispatching events for when expanded/collapsed - good to do (could be the one event) `{detail: { opened: true } }` - should bubble and not be cancelable. `new CustomEvent('wagtail:panel-toggle', { bubbles: true, cancelable: false, detail: { opened: true } }),`
* Adding event listener for triggering this from other code - probably not critical, custom JS could just trigger a 'click' if really needed.
* Potential for an event or some global approach to instantiate custom panels - not critical for 4.0 as use case for async HTML injection of panels is minimal (e.g. not really in modals or listings).